### PR TITLE
Add pykafka recipe

### PR DESCRIPTION
**Pull request by mrocklin**
cc @emmett9001

Depends on #4090 

See https://github.com/Parsely/pykafka/issues/730

### CODE REVIEW

**Comment by mrocklin** [comment]
@emmett9001 I found that I needed to install gevent in order to import pykafka.  Is this expected?

File: recipes/pykafka/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+      PyKafka’s primary goal is to provide a similar level of abstraction to
+      the JVM Kafka client using idioms familiar to Python programmers and
+      exposing the most Pythonic API possible.
+  doc_url: http://pykafka.readthedocs.org/
+  dev_url: https://github.com/parsely/pykafka
+
+extra:
+  recipe-maintainers:
+    - mrocklin
+    - emmett9001
```
Comment (jakirkham) [inline recipes/pykafka/meta.yaml]: Are you ok being a maintainer, @emmett9001?

**Comment by jakirkham** [comment]
Just want to check with the other person to make sure they are ok maintaining this. Otherwise LGTM.

**Comment by mrocklin** [comment]
> Just want to check with the other person to make sure they are ok maintaining this. Otherwise LGTM.

Yes, we're in conversation in https://github.com/Parsely/pykafka/issues/730 .  I haven't yet received a confirmation.

Also, what is the right thing to do if the package needs `pkg_resources` at runtime?  Add a runtime dependency on setuptools?

**Comment by mrocklin** [comment]
It looks like we're not picking up the librdkafka library at build time.  http://pykafka.readthedocs.io/en/latest/#using-the-librdkafka-extension

Is there a nice way to add conda's local include and lib directories to `C_INCLUDE_PATH` and `LD_LIBRARY_PATH`?  Is this typically done in a build.sh script or is there some more standard approach?

File: recipes/pykafka/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  sha256: {{ sha256 }}
+
+build:
+  noarch: python
+  number: 0
+  script: python setup.py install --single-version-externally-managed --record record.txt
+
+requirements:
+  build:
+    - python
```
Comment (jakirkham) [inline recipes/pykafka/meta.yaml]: Please add `toolchain`.

**Comment by jakirkham** [comment]
> Also, what is the right thing to do if the package needs `pkg_resources` at runtime?  Add a runtime dependency on setuptools?

Yes.

> Is there a nice way to add conda's local include and lib directories to `C_INCLUDE_PATH` and `LD_LIBRARY_PATH`?  Is this typically done in a build.sh script or is there some more standard approach?

Adding `toolchain` above should fix it.

**Comment by mrocklin** [comment]
This recipe has grown somewhat more complex.  Does it still meet the requirements of `noarch: python`?

**Comment by mrocklin** [comment]
It looks like we're still not able to see the librdkafka library.  I'm not sure if this is at build time or at runtime though.

**Comment by mrocklin** [comment]
Failing at build time

    pykafka/rdkafka/_rd_kafkamodule.c:11:32: fatal error: librdkafka/rdkafka.h: No such file or directory

**Comment by jakirkham** [comment]
> Does it still meet the requirements of `noarch: python`?

If it is compiling stuff, would say no.

File: recipes/pykafka/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    - setuptools
+    - toolchain
```
Comment (jakirkham) [inline recipes/pykafka/meta.yaml]: Please add `librdkafka` here too if it is needed for the build.

Comment (mrocklin) [reply recipes/pykafka/meta.yaml]: Ah ah!  Of course.  Now I feel a bit silly.  Thanks @jakirkham !

Comment (jakirkham) [reply recipes/pykafka/meta.yaml]: np

File: recipes/pykafka/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+build:
+  number: 0
+  script: python setup.py install --single-version-externally-managed --record record.txt
+
+requirements:
+  build:
+    - python
+    - setuptools
+    - toolchain
+    - librdkafka  # [not win]
```
Comment (jakirkham) [inline recipes/pykafka/meta.yaml]: Normally for libraries that are linked to, we pick some sort of pinning so the stack remains consistent and things don't break when a new version comes out. Based on the [API/ABI report]( https://abi-laboratory.pro/tracker/timeline/librdkafka/ ), it looks like `librdkafka` breaks on patch versions sometimes. Though `0.9.4` and `0.9.5` seem ok. Given this, it seems reasonable to change this to `librdkafka 0.9.4|0.9.5`. Thoughts on this?

cc @conda-forge/librdkafka

Comment (kwilcox) [reply recipes/pykafka/meta.yaml]: :+1: I'm pinned at `0.9.4` here; this works for me.

I'd argue it's important to keep all of the libraries that depend on `librdfakfa` on the same pin so they are interchangable. Is there a non-manual way to do this other than a global `conda-forge` pin?

*  https://github.com/conda-forge/python-confluent-kafka-feedstock
*  https://github.com/conda-forge/kafkacat-feedstock

Comment (mrocklin) [reply recipes/pykafka/meta.yaml]: Hrm, at runtime I would expect conda itself to figure this out.  I'm not sure how feasible it is to use pykafka with a different librdkafka from which it was built.

File: recipes/pykafka/meta.yaml (hunk 1 above)
Comment (emmett9001) [reply recipes/pykafka/meta.yaml]: Yes I am!

**Comment by emmett9001** [comment]
@mrocklin gevent is still required as of 2.6.0, but 2.7.0.dev1 [removes](https://github.com/Parsely/pykafka/pull/686) the dependency.

**Comment by emmett9001** [comment]
How does conda-forge handle development versions? I've been doing pre-release versions for a while (with `.dev` on the end of the version number), and I know setuptools automatically ignores these versions during `pip install`. Moving forward, should I also do conda-forge pre-release versions this way?

File: recipes/pykafka/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,60 @@
+{% set name = "pykafka" %}
+{% set version = "2.6.0" %}
+{% set sha256 = "19e4a802b00e28c3203fcfd80ca440b13faecb8e58171e9685a044cf87dae93e" %}
```
[the remark this answers is not in the document]
Comment (mrocklin) [reply recipes/pykafka/meta.yaml]: https://pypi.org/project/pykafka/#files

**Comment by mrocklin** [comment]
> How does conda-forge handle development versions? I've been doing pre-release versions for a while (with .dev on the end of the version number), and I know setuptools automatically ignores these versions during pip install. Moving forward, should I also do conda-forge pre-release versions this way?

I honestly never do dev releases on conda-forge.  But then again I rarely do dev releases.  I don't have much to provide here.  @jakirkham generally knows best practices.

> @mrocklin gevent is still required as of 2.6.0, but 2.7.0.dev1 removes the dependency.

OK, good to know.

From my perspective this PR is good to go.

**Comment by mrocklin** [comment]
Is there anything else here that needs to be resolved?

**Comment by mrocklin** [comment]
Ping

File: recipes/pykafka/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+build:
+  number: 0
+  script: python setup.py install --single-version-externally-managed --record record.txt
+
+requirements:
+  build:
+    - python
+    - setuptools
+    - toolchain
+    - librdkafka 0.9.4|0.9.5  # [not win]
```
Comment (jakirkham) [inline recipes/pykafka/meta.yaml]: Could you please add this pinning to [our list of pinnings]( https://github.com/conda-forge/conda-forge.github.io/blob/master/scripts/pin_the_slow_way.py#L40 )?

**Comment by jakirkham** [comment]
Sorry @mrocklin. Got a bit swamped.

Would like us to add the `librdkafka` pinning to our list of pinnings. Otherwise this should be good to go.

**Comment by mrocklin** [comment]
> Would like us to add the librdkafka pinning to our list of pinnings. Otherwise this should be good to go.

I'm not sure I understand the intent here or the format of that document.  I'm happy to submit a PR that follows that pattern though.  What happens if we update librdkafka on conda-forge (which is a few versions old at the moment).

**Comment by jakirkham** [comment]
So we try to keep all pinnings the same throughout conda-forge. That way, one can confidently install and use various packages from conda-forge without running into dependency conflicts.

ATM we pin `build` and `run` the same (though that may change in the future). The package name is listed on the left and on the right is the pinning that we are using in the org. The comment is there to note what `defaults` is pinning a package to (if known or if it exists in `defaults`).

Hope that helps. If you have more questions, please feel free to ask. :)

**Comment by mrocklin** [comment]
Ah, so all conda-forge recipes are pinned to the exact same version?  How/when does the community decide to bump these pinned versions?

**Comment by jakirkham** [comment]
Yeah we try our best to do so. Though we haven't equipped the linter to enforce this and some recipes lag (though the pinning bot should help here). So there could be some exceptions. That said, the goal is to keep everything in lockstep.

Normally someone in the community feels a need to have a newer version and they PR a version bump to the webpage repo. Then we try to discuss with relevant parties to see if that works for everyone affected. Though we may go ahead if some people are slow to reply.

**Comment by jakirkham** [comment]
Thanks @mrocklin.

**Comment by mrocklin** [comment]
Thank you for shepherding this along @jakirkham

**Pull request merged**